### PR TITLE
Pre-commit implementation

### LINKED ISSUE #30265: Plan implemention of pre-commit hooks using pre-commit framework
Use [pre-commit](https://pre-commit.com/) framework to add hooks for checking various aspects of the change set before it is committed. This should enable fewer round trips to the CI server when the initial static analysis checks fail.

Before:
* Write suggestion document about how to implement, how developers will setup, etc
* Yapf format code PR + CI integration
* Upgrade clang-format version to latest PR + CI
* Think about manually building each hook (Ensuring that the version numbers are the exact same as CI with one update)

In order:
* check-xml
* check-yaml
* Clang-format in place (Maintain self using docker in seperate repo)
* Yapf (Using current yapf style)
* Flake8

In future:
* no-commit-to-branch
* trailing-whitespace
* Cppcheck
* + More suggestions

We could start with clang-format and add more along the way. I experimented with static clang-format pre-commit hook that came with the correct clang-format version for each platform at https://github.com/martyngigg/pre-commit-clang-format.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
Implement a config for pre-commit and fix the clang-formatting issues on all files.

**To test:**
Run this command in a git bash/terminal to see it in action on the whole repo:
`pre-commit run --all-files`

Fixes #30265 

*No release notes as not a user facing change*

The version into `ornl-next` is #30981.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  # Flake8 isn't automatically parallel on Windows due to lack of multiprocessing library support. So default to 16 jobs
+  # to support 8 hyper threaded cores, In practice the drop on other OS' should be negligible
+  - repo: https://gitlab.com/pycqa/flake8
+    rev: 3.7.9
+    hooks:
+      - id: flake8
+        # Have to manually copy this from the .flake8 as flake8 will just ignore the config if the file is passed
+        exclude: build|buildconfig|docs|Framework/Algorithms/test|Framework/DataObjects|Framework/MPIAlgorithms|Framework/PythonInterface/mantid|Framework/PythonInterface/test|images|installers|instrument|MantidPlot|qt/applications/workbench/workbench/app/resources.py|qt/paraview_ext|qt/python/mantidqt/resources.py|qt/python/mantidqt/dialogs/errorreports/resources_qt4.py|qt/python/mantidqt/dialogs/errorreports/resources_qt5.py|qt/scientific_interfaces|qt/widgets|Testing/PerformanceTests|Testing/SystemTests/lib|Testing/SystemTests/scripts|Testing/SystemTests/tests/framework/reference|Testing/Tools|scripts/ErrorReporter/ui_errorreport.py|scripts/Interface/ui|scripts/Interface/ui/dataprocessorinterface
+        args: ['--extend-ignore=E999,F821,E127', '-j16']
```
Comment (peterfpeterson) [inline .pre-commit-config.yaml:34]: Just because my editor will add the line if I go near this file, please add EOL to the last line.

Comment (Pasarus) [reply .pre-commit-config.yaml]: Done :+1:

File: buildconfig/CMake/CommonSetup.cmake [hunk 2]
```diff
@@ -412,6 +412,23 @@ if(NOT BUNDLES)
   set(BUNDLES "./")
 endif()
 
+# ##############################################################################
+# Setup pre-commit here as otherwise it will be overwritten by earlier 
+# pre-commit hooks being added
+# ##############################################################################
+if (MSVC)
```
Comment (peterfpeterson) [inline buildconfig/CMake/CommonSetup.cmake:419]: Please put this inside an `ENABLE_PRECOMMIT` which can be setup with a line before it being
```cmake
option(ENABLE_PRECOMMIT "Enable pre-commit framework" ON)
```

Comment (Pasarus) [reply buildconfig/CMake/CommonSetup.cmake]: Done :+1:

File: buildconfig/CMake/CommonSetup.cmake [hunk 3]
```diff
@@ -412,6 +412,27 @@ if(NOT BUNDLES)
...
+    # Create pre-commit script wrapper to use mantid third party python for pre-commit
+    execute_process(COMMAND bash -c "(cd ${PROJECT_SOURCE_DIR}; mv .git/hooks/pre-commit .git/hooks/pre-commit-script)")
+    execute_process(COMMAND bash -c "(cd ${PROJECT_SOURCE_DIR}; echo '#!/bin/sh\n${MSVC_PYTHON_EXECUTABLE_DIR}/python.exe ${PROJECT_SOURCE_DIR}/.git/hooks/pre-commit-script' >> .git/hooks/pre-commit)")
+  else()
+    # Use system installed pre-commit if not present it should just fail but 
+    # continue anyway.
+    execute_process(COMMAND bash -c "(cd ${PROJECT_SOURCE_DIR}; pre-commit install)")
+  endif()
+else()
+  message(AUTHOR_WARNING "Pre-commit not enabled by CMake, please enable manually.")
```
Comment (peterfpeterson) [inline buildconfig/CMake/CommonSetup.cmake:433]: I don't know under what condition this would show.

Comment (Pasarus) [reply buildconfig/CMake/CommonSetup.cmake:433]: I don't think there is as this code runs before the user has a chance to configure i.e. before clicking generate. Not sure if it is worth having the Pre_commit option?

Comment (Pasarus) [reply buildconfig/CMake/CommonSetup.cmake:433]: Just to clarify on the first run in the CMake GUI it will not trigger, but it will on the second time you click configure, or if a developer passed -DENABLE_PRECOMMIT=OFF it would show this warning.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is looking really good and I think we're nearly there.

One thing to consider though is I think it might be an idea to add something about the fact we use `pre-commit` and what it is in the dev docs GitWorkflow page? I'm thinking of new developers that won't know what it is or what it is for. A brief paragraph and links to the pre-commit docs would be fine I think.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply .pre-commit-config.yaml:38]: It's no issue to have multiple threads that aren't directly on a core/logical hyper-threaded core. As the OS will handle it's multi-threading, regardless of numbers, without a problem.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    hooks:
+      - id: clang-format
+        exclude: Testing/Tools/cxxtest|tools
+
+  - repo: https://github.com/google/yapf
+    rev: a53e340c2ff24b36c98945c6f0374c6e21899afe
+    hooks:
+      - id: yapf
+
+  # Flake8 isn't automatically parallel on Windows due to lack of multiprocessing library support. So default to 16 jobs
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply .pre-commit-config.yaml:30]: It does not I will remove it.

File: buildconfig/Jenkins/pre-commit [hunk 6]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+# Check that the source conforms to the pre-commit framework
+# Args are expected as GIT_SHA1 as $1, PROJECT_SOURCE as $2, JOB_TYPE as $3
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply buildconfig/Jenkins/pre-commit:3]: Will do :+1:

File: buildconfig/Jenkins/pre-commit [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  echo "Usage: ./pre-commit branch-sha1 project-root job-type"
+  echo "  job_type=master|pull-request  "
+  exit 1
+fi
+
+GIT_SHA=$1
+PROJECT_SOURCE=$2
+JOB_TYPE=$3
+
+cd PROJECT_SOURCE
```
Comment (Pasarus) [inline buildconfig/Jenkins/pre-commit:15]:
```suggestion
cd $PROJECT_SOURCE
```

File: buildconfig/Jenkins/pre-commit [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+  PRE_COMMIT_VERSION=$(pre-commit --version)
+  # If using the snap version (No longer updated but it doesn't use pip on ubuntu, so it's better to be outdated here) run
+  # Using the older source
+  if [[ ${PRE_COMMIT_VERSION} < "pre-commit 1.22.0" ]]; then
+    SCRIPT_OUTPUT=$(pre-commit run --origin ${BASE_SHA} --source ${GIT_SHA} --show-diff-on-failure --verbose)
+  else
+    SCRIPT_OUTPUT=$(pre-commit run --from-ref ${BASE_SHA} --to-ref ${GIT_SHA} --show-diff-on-failure --verbose)
+  fi
+elif [[ JOB_TYPE == "master" ]]; then
```
Comment (Pasarus) [inline buildconfig/Jenkins/pre-commit:29]:
```suggestion
elif [[ $JOB_TYPE == "master" ]]; then
```

File: buildconfig/Jenkins/pre-commit [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  exit 1
+fi
+
+GIT_SHA=$1
+PROJECT_SOURCE=$2
+JOB_TYPE=$3
+
+cd $PROJECT_SOURCE
+
+if [[ JOB_TYPE == "pull-request" ]]; then
```
Comment (Pasarus) [inline buildconfig/Jenkins/pre-commit:17]:
```suggestion
if [[ $JOB_TYPE == "pull-request" ]]; then
```

File: buildconfig/CMake/CommonSetup.cmake [hunk 10]
```diff
@@ -412,6 +402,30 @@ if(NOT BUNDLES)
...
+# ##############################################################################
+# Setup pre-commit here as otherwise it will be overwritten by earlier
+# pre-commit hooks being added
+# ##############################################################################
+option(ENABLE_PRECOMMIT "Enable pre-commit framework" ON)
+if (ENABLE_PRECOMMIT)
+  if (MSVC)
+    # Use downloaded ThirdParty version of pre-commit
+    execute_process(COMMAND bash -c "${MSVC_PYTHON_EXECUTABLE_DIR}/Scripts/pre-commit.cmd install" WORKING_DIRECTORY ${PROJECT_SOURCE_DIR})
+    # Create pre-commit script wrapper to use mantid third party python for pre-commit
```
Comment (martyngigg) [inline buildconfig/CMake/CommonSetup.cmake:414]: I think it would be good to replace these next 2 `execute_process` commands and use [`file`](https://cmake.org/cmake/help/v3.15/command/file.html?highlight=file) instead to avoid starting 2 new subprocesses that are particularly expensive on windows.

File: buildconfig/Jenkins/pre-commit [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+if [ $# -ne 3 ]; then
+  echo "Usage: ./pre-commit branch-sha1 project-root job-type"
+  echo "  job_type=master|pull-request  "
+  exit 1
+fi
+
+GIT_SHA=$1
+PROJECT_SOURCE=$2
+JOB_TYPE=$3
```
Comment (peterfpeterson) [inline buildconfig/Jenkins/pre-commit:13]: Is this is `JOB_NAME` in other buildscripts?

Comment (Pasarus) [reply buildconfig/Jenkins/pre-commit:13]: So JOB_TYPE is presently just passed into the script, it's defined by whomever calls the script, it's a kinda bespoke method for this script, so you pass master for your main branch or pull-request for just the diff.

Comment (peterfpeterson) [reply buildconfig/Jenkins/pre-commit:13]: Does this mean `buildscript` and `buildscript.bat` need to be modified to call this script?

Comment (Pasarus) [reply buildconfig/Jenkins/pre-commit:13]: Why would they need to call this script?

Comment (martyngigg) [reply buildconfig/Jenkins/pre-commit:13]: @peterfpeterson We decided to just have it disabled in the main buildscript using the above option as we'll have separate jobs for them with their own scripts.

**Review by martyngigg — DISMISSED** [review]
I'm happy with how this looks now.

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    rev: 2618198e9658cccb4a53f04af0f7d642109f3b54
+    hooks:
+      - id: trailing-whitespace
+        args: [--markdown-linebreak-ext=md]
+      - id: check-added-large-files
+        args: ['--maxkb=4096']
+      - id: check-xml
+      - id: check-yaml
+
+  - repo: https://github.com/mantidproject/pre-commit-hooks
```
Comment (peterfpeterson) [inline .pre-commit-config.yaml:19]: clang-format and yapf should be done in separate PRs as they have knock-on effects for changing lots of other files. Even not on the build server, developers will be changing a ton of files unnecessarily.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 12 above)
Comment (peterfpeterson) [reply .pre-commit-config.yaml:19]: Otherwise, I'm happy with what is proposed here. I am just concerned that the "normal" developer will make a bunch of conflicts.

Comment (martyngigg) [reply .pre-commit-config.yaml:19]: Yep. Totally agree. We were going to wait until the last minute to open that PR before we would then merge this.

**Comment by martyngigg** [comment]
@peterfpeterson We've dropped all of the other formatting stuff here so this should be ready to go in.

**Comment by martyngigg** [comment]
Just to confirm. The config now has the basic checks that are not covered by one of our other CI jobs and this included trimming trailing whitespace. The changes from that are included here.

**Comment by martyngigg** [comment]
The macOS failure is unrelated. I've restarted it.

**Comment by martyngigg** [comment]
@peterfpeterson How happy would you be to approve this now?

**Review by peterfpeterson — APPROVED** [review]
